File: src/python/sequential/graph/bfs/breadth_first_traversal.py
import collections


def bfs(graph, root):
    """
    Performs breadth-first traversal on the :param graph: starting from the :param root: node.
    """
    explored = set()
    traversed = []
    search_queue = collections.deque()

    search_queue.append(root)
    while search_queue:
        node = search_queue.popleft()
        if node not in explored:
            explored.add(node)
            traversed.append(node)
            search_queue += graph.get(node, [])

    return traversed


def main():
    graph = {
        "you": ["alice", "bob", "clair"],
        "alice": ["peggy"],
        "clair": ["thom", "jonny"],
        "bob": ["anuj", "peggy"],
    }

    print(bfs(graph, "you"))

if __name__ == "__main__":
    main()

File: src/swift/sequential/graph/bfs/BreadthFirstTraversal.swift
typealias Graph<T: Hashable> = [T: [T]]


func bfs<T>(in graph: Graph<T>, withRoot root: T) -> [T] {
    var explored = Set<T>()
    var traversed = [T]()
    var searchQueue = [root]
    var queuePopIndex = 0
    
    while queuePopIndex < searchQueue.count {
        let node = searchQueue[queuePopIndex]
        queuePopIndex += 1
        
        if !explored.contains(node) {
            explored.insert(node)
            traversed.append(node)
            if let successors = graph[node] {
                searchQueue += successors
            }
        }
    }
    return traversed
}


func mainBfs() {
    let graph = [
        "you": ["alice", "bob", "clair"],
        "alice": ["peggy"],
        "clair": ["thom", "jonny"],
        "bob": ["anuj", "peggy"]
    ]

    print(
        bfs(in: graph, withRoot: "you")
    )
}